src/CompileCommand.cpp

```diff
@@ -0,0 +1,335 @@
+//
+// cat include/clang/Driver/Options.td |
+//   grep -E 'JoinedOrSeparate.*(CC1Option|(Preprocessor|M|clang_i|I)_Group)'
+//
+// This function duplicates this behavior.
+
+ArgStatus
+inspect_argument(string_view arg)
+{
+  // need at least "-M" to be removed or "-DX" to be split and start with a "-"
+  if (arg.length() < 2 || arg.at(0) != '-') {
+    return ArgStatus::keep();
+  }
+
+  char second = arg.at(1);
+
+  auto keep_and_maybe_split = [](string_view arg1,
+                                 const char* arg2) -> ArgStatus {
+    if (arg1 == arg2) {
+      return ArgStatus::keep();
+    } else {
+      return ArgStatus::keep_and_split_at(string_view(arg2).length());
+    }
+  };
+
+  // single character options
+  if (second == 'D' || second == 'F' || second == 'I' || second == 'U'
+      || second == 'o') {
+    if (arg.length() > 2) {
+      return ArgStatus::keep_and_split_at(2);
+    } else {
+      return ArgStatus::keep();
+    }
+  } else if (second == 'x') { // ["-x", "c++"] or ["-xc++"]
+    if (arg.length() == 2) {
+      return ArgStatus::skip_arg_also();
+    } else {
+      return ArgStatus::skip();
+    }
+  }
+  // -M options: skip all of them
+  else if (second == 'M') {
+    if (arg.length() == 2) {
+      return ArgStatus::skip();
+    }
+    char third = arg.at(2);
+    switch (third) {
+    case 'D': // -MD
+    case 'G': // -MG
+    case 'P': // -MP
+    case 'V': // -MV
+      return ArgStatus::skip();
+
+    case 'M': { // -MM or // -MMD
+      if (arg == "-MM" || arg == "-MMD") {
+        return ArgStatus::skip();
+      } else {
+        // unknown -MM?
+        return ArgStatus::unknown();
+      }
+    }
+
+    // these can be provided as ["-MF", "arg"] or ["-MFarg"], skip both
+    case 'F':
+    case 'J':
+    case 'Q':
+    case 'T':
+      return (arg.length() == 3) ? ArgStatus::skip_arg_also()
+                                 : ArgStatus::skip();
+
+    default:
+      // lowercase next, e.g. -Mach
+      if (third >= 'a' || third <= 'z') {
+        return ArgStatus::keep();
+      }
+      // other unknown -M? arg
+      return ArgStatus::unknown();
+    }
+  }
+  // remaining arguments
+  else if (second == 'i') {
+    const char* i_args[] = {
+      "-interface-stub-version=",
+      "-idirafter",
+      "-iframework",
+      "-imacros",
+      "--imacros",
+      "-include",
+      "--include",
+      "-iprefix",
+      "-iquote",
+      "-isysroot",
+      "-isystem",
+      "-iwithprefixbefore",
+      "-iwithprefix",
+      "-iwithsysroot",
+      "-ivfsoverlay",
+    };
+    for (const char* i_arg : i_args) {
+      if (arg.starts_with(i_arg)) {
+        return keep_and_maybe_split(arg, i_arg);
+      }
+    }
+  } else {
+    for (const char* misc_arg : {"-cxx-isystem", "-working-directory"}) {
+      if (second == misc_arg[1] && arg.starts_with(misc_arg)) {
+        return keep_and_maybe_split(arg, misc_arg);
+      }
+    }
+  }
+
+  return ArgStatus::keep();
+}
+
+size_t
```

Comment:
What if the call is "ccache distcc gcc", then not distcc but json should be in the compile commands file.